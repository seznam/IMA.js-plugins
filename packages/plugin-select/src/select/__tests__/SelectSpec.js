import { PageStateManager, Dispatcher, PageContext } from '@ima/core';
import { shallow, mount } from 'enzyme';
import React from 'react';
import { toMockedInstance, setGlobalMockMethod } from 'to-mock';
import forwardedSelect, {
  createStateSelector,
  select,
  setCreatorOfStateSelector,
  setHoistStaticMethod,
  hoistNonReactStatic
} from '../select';

setGlobalMockMethod(jest.fn);

describe('plugin-select:', () => {
  const appState = {
    media: {
      width: 90,
      height: 60
    },
    title: 'title'
  };
  const componentContext = {
    $Utils: {
      $PageStateManager: toMockedInstance(PageStateManager, {
        getState: () => {
          return appState;
        }
      }),
      $Dispatcher: toMockedInstance(Dispatcher)
    }
  };
  const selectorMethods = [
    state => {
      return {
        width: state.media.width
      };
    },
    state => {
      return {
        height: state.media.height
      };
    }
  ];

  const selectorUsingProps = (state, context, props) => ({
    width: state.media.width * props.multiplier,
    height: state.media.height * props.multiplier
  });

  beforeEach(() => {
    global.$Debug = true;
  });

  afterEach(() => {
    delete global.$Debug;
  });

  describe('createStateSelector', () => {
    it('should select extra properties from state', () => {
      let extraProps = createStateSelector(...selectorMethods)(
        appState,
        componentContext
      );

      expect(extraProps).toMatchSnapshot();
    });

    it('should return same output for same input', () => {
      let selector = createStateSelector(...selectorMethods);

      let obj1 = selector(appState, componentContext);
      let obj2 = selector(appState, componentContext);

      expect(obj1 === obj2).toBeTruthy();
    });

    it('should trow error for undefined $PageStateManager', () => {
      expect(() => {
        createStateSelector(...selectorMethods)({ context: { $Utils: {} } });
      }).toThrow();
    });
  });

  describe('select', () => {
    let wrapper = null;
    const defaultProps = {
      props: 'props',
      multiplier: 0.5
    };

    class Component extends React.PureComponent {
      static defaultProps() {
        return defaultProps;
      }

      render() {
        return <h1>text</h1>;
      }
    }

    const MockContextProvider = ({ children }) => (
      <PageContext.Provider value={componentContext}>
        {children}
      </PageContext.Provider>
    );

    it('should render component', () => {
      wrapper = shallow(React.createElement(Component, defaultProps), {
        context: componentContext
      });

      expect(wrapper).toMatchSnapshot();
    });

    it('should render component with extraProps', () => {
      let EnhancedComponent = select(...selectorMethods)(Component);

      wrapper = mount(React.createElement(EnhancedComponent, defaultProps), {
        context: componentContext,
        wrappingComponent: MockContextProvider
      });

      expect(wrapper).toMatchSnapshot();
    });

    it('should render component with extraProps modifies by ownProps', () => {
      let EnhancedComponent = select(
        ...selectorMethods,
        selectorUsingProps
      )(Component);

      wrapper = mount(React.createElement(EnhancedComponent, defaultProps), {
        context: componentContext,
        wrappingComponent: MockContextProvider
      });

      expect(wrapper).toMatchSnapshot();
    });

    it('should add listener to dispatcher after mounting to DOM', () => {
      let EnhancedComponent = select(...selectorMethods)(Component);

      wrapper = mount(React.createElement(EnhancedComponent, defaultProps), {
        context: componentContext,
        wrappingComponent: MockContextProvider
      });

      expect(componentContext.$Utils.$Dispatcher.listen).toHaveBeenCalled();
    });

    it('should remove listener to dispatcher before unmounting from DOM', () => {
      let EnhancedComponent = select(...selectorMethods)(Component);

      wrapper = mount(React.createElement(EnhancedComponent, defaultProps), {
        context: componentContext,
        wrappingComponent: MockContextProvider
      });

      wrapper.unmount();

      expect(componentContext.$Utils.$Dispatcher.unlisten).toHaveBeenCalled();
    });

    it('should render component with extraProps and own createStateSelector', () => {
      setCreatorOfStateSelector((...selectors) => {
        return (state, context) => {
          return selectors.reduce((result, selector) => {
            return Object.assign({}, result, selector(state, context));
          }, {});
        };
      });
      let EnhancedComponent = select(...selectorMethods)(Component);

      wrapper = mount(React.createElement(EnhancedComponent, defaultProps), {
        context: componentContext,
        wrappingComponent: MockContextProvider
      });

      expect(wrapper).toMatchSnapshot();
    });

    it('should render component with extraProps and own static methods', () => {
      setHoistStaticMethod((TargetComponent, Original) => {
        const keys = Object.getOwnPropertyNames(Original);

        keys.forEach(key => {
          if (key === 'defaultProps') {
            const descriptor = Object.getOwnPropertyDescriptor(Original, key);
            try {
              Object.defineProperty(TargetComponent, key, descriptor);
            } catch (e) {} // eslint-disable-line no-empty
          }
        });

        return hoistNonReactStatic(TargetComponent, Original);
      });
      let EnhancedComponent = select(...selectorMethods)(Component);

      wrapper = mount(React.createElement(EnhancedComponent, defaultProps), {
        context: componentContext,
        wrappingComponent: MockContextProvider
      });

      expect(typeof EnhancedComponent.defaultProps === 'function').toBeTruthy();
      expect(wrapper).toMatchSnapshot();
    });

    it('should forward ref', () => {
      let EnhancedComponent = forwardedSelect(...selectorMethods)(Component);
      let props = Object.assign({}, defaultProps, {
        ref: React.createRef()
      });

      wrapper = shallow(React.createElement(EnhancedComponent, props), {
        context: componentContext
      });

      expect(wrapper).toMatchSnapshot();
    });
  });
});
