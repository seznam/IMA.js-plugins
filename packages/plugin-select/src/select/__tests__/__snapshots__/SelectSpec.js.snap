// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin-select: createStateSelector should return same output for dynamic changing state selector keys with same values 1`] = `
Object {
  "height": 60,
  "width": 90,
}
`;

exports[`plugin-select: createStateSelector should return same output for dynamic changing state selector keys with same values 2`] = `
Object {
  "dynamic": 1,
  "height": 60,
  "width": 90,
}
`;

exports[`plugin-select: createStateSelector should return same output for dynamic changing state selector keys with same values 3`] = `
Object {
  "dynamic": 1,
  "height": 60,
  "width": 90,
}
`;

exports[`plugin-select: createStateSelector should return same output for dynamic changing state selector keys with same values 4`] = `
Object {
  "height": 60,
  "width": 90,
}
`;

exports[`plugin-select: createStateSelector should select extra properties from state 1`] = `
Object {
  "height": 60,
  "width": 90,
}
`;

exports[`plugin-select: select should forward ref 1`] = `
<withContext(Component)
  forwardedRef={
    Object {
      "current": null,
    }
  }
  multiplier={0.5}
  props="props"
  settings={
    Object {
      "color": "red",
    }
  }
/>
`;

exports[`plugin-select: select should render component 1`] = `
<h1>
  text
</h1>
`;

exports[`plugin-select: select should render component with extraProps 1`] = `
<withContext(Component)
  multiplier={0.5}
  props="props"
  settings={
    Object {
      "color": "red",
    }
  }
>
  <SelectState
    $context={
      Object {
        "$Utils": Object {
          "$Dispatcher": Mock {},
          "$PageStateManager": Mock {
            "getState": [Function],
          },
        },
      }
    }
    multiplier={0.5}
    props="props"
    settings={
      Object {
        "color": "red",
      }
    }
  >
    <Component
      $context={
        Object {
          "$Utils": Object {
            "$Dispatcher": Mock {},
            "$PageStateManager": Mock {
              "getState": [Function],
            },
          },
        }
      }
      height={60}
      multiplier={0.5}
      props="props"
      settings={
        Object {
          "color": "red",
        }
      }
      width={90}
    >
      <h1>
        text
      </h1>
    </Component>
  </SelectState>
</withContext(Component)>
`;

exports[`plugin-select: select should render component with extraProps and own createStateSelector 1`] = `
<withContext(Component)
  multiplier={0.5}
  props="props"
  settings={
    Object {
      "color": "red",
    }
  }
>
  <SelectState
    $context={
      Object {
        "$Utils": Object {
          "$Dispatcher": Mock {},
          "$PageStateManager": Mock {
            "getState": [Function],
          },
        },
      }
    }
    multiplier={0.5}
    props="props"
    settings={
      Object {
        "color": "red",
      }
    }
  >
    <Component
      $context={
        Object {
          "$Utils": Object {
            "$Dispatcher": Mock {},
            "$PageStateManager": Mock {
              "getState": [Function],
            },
          },
        }
      }
      height={60}
      multiplier={0.5}
      props="props"
      settings={
        Object {
          "color": "red",
        }
      }
      width={90}
    >
      <h1>
        text
      </h1>
    </Component>
  </SelectState>
</withContext(Component)>
`;

exports[`plugin-select: select should render component with extraProps and own static methods 1`] = `
<withContext(Component)
  multiplier={0.5}
  props="props"
  settings={
    Object {
      "color": "red",
    }
  }
>
  <SelectState
    $context={
      Object {
        "$Utils": Object {
          "$Dispatcher": Mock {},
          "$PageStateManager": Mock {
            "getState": [Function],
          },
        },
      }
    }
    multiplier={0.5}
    props="props"
    settings={
      Object {
        "color": "red",
      }
    }
  >
    <Component
      $context={
        Object {
          "$Utils": Object {
            "$Dispatcher": Mock {},
            "$PageStateManager": Mock {
              "getState": [Function],
            },
          },
        }
      }
      height={60}
      multiplier={0.5}
      props="props"
      settings={
        Object {
          "color": "red",
        }
      }
      width={90}
    >
      <h1>
        text
      </h1>
    </Component>
  </SelectState>
</withContext(Component)>
`;

exports[`plugin-select: select should render component with extraProps modifies by ownProps 1`] = `
<withContext(Component)
  multiplier={0.5}
  props="props"
  settings={
    Object {
      "color": "red",
    }
  }
>
  <SelectState
    $context={
      Object {
        "$Utils": Object {
          "$Dispatcher": Mock {},
          "$PageStateManager": Mock {
            "getState": [Function],
          },
        },
      }
    }
    multiplier={0.5}
    props="props"
    settings={
      Object {
        "color": "red",
      }
    }
  >
    <Component
      $context={
        Object {
          "$Utils": Object {
            "$Dispatcher": Mock {},
            "$PageStateManager": Mock {
              "getState": [Function],
            },
          },
        }
      }
      height={30}
      multiplier={0.5}
      props="props"
      settings={
        Object {
          "color": "red",
        }
      }
      width={45}
    >
      <h1>
        text
      </h1>
    </Component>
  </SelectState>
</withContext(Component)>
`;

exports[`plugin-select: select should render component with extraProps replaced by ownProps 1`] = `
<withContext(Component)
  multiplier={0.5}
  props="props"
  settings={
    Object {
      "color": "red",
    }
  }
>
  <SelectState
    $context={
      Object {
        "$Utils": Object {
          "$Dispatcher": Mock {},
          "$PageStateManager": Mock {
            "getState": [Function],
          },
        },
      }
    }
    multiplier={0.5}
    props="props"
    settings={
      Object {
        "color": "red",
      }
    }
  >
    <Component
      $context={
        Object {
          "$Utils": Object {
            "$Dispatcher": Mock {},
            "$PageStateManager": Mock {
              "getState": [Function],
            },
          },
        }
      }
      height={60}
      multiplier={0.25}
      props="props"
      settings={
        Object {
          "color": "red",
          "newSettingsProp": true,
        }
      }
      width={90}
    >
      <h1>
        text
      </h1>
    </Component>
  </SelectState>
</withContext(Component)>
`;
